### ISSUE TITLE
Any plan to support draft 6 ?

### ISSUE THREAD

**Issue by Airkro** [json-schema-faker/json-schema-faker#289]
It will be great

**Comment by pateketrueke** [comment]
As soon a stable `0.5.x` version is released I think we might to consider it, but afaik it can be more complex to build by myself.

Thanks for your feedback!

**Comment by ehmicky** [comment]
Some aspects of JSON schema v6 and v7 are complex, but some might be low-hanging fruits. Libraries like [`ajv`](https://github.com/epoberezkin/ajv) often only support parts of the most recent properties.

If you planned to support it, this is a list of the things that might change or be added:
* JSON schemas can now be booleans. Not sure how `false` should be supported, but `true` could re-use the `anyValue()` function
* [`id`](https://github.com/json-schema-faker/json-schema-faker/blob/master/ts/core/run.ts#L101) is renamed `$id`
* The `isKey()` utility should also exclude the following properties: `examples`
* New formats, i.e. maybe new RegExps in [`coreFormat`](https://github.com/json-schema-faker/json-schema-faker/blob/master/ts/generators/coreFormat.ts): `regex`, `idn-email`, `idn-hostname`, `uri-template`, `iri`, `iri-reference`, `json-pointer`, `relative-json-pointer`. You can probably copy the RegExps from libraries like ajv.
* New `const` keyword: should be like `enum` but with single possibility
* `exclusiveMaximum` and `exclusiveMinimum` are now numbers not booleans
* New keywords that might be harder to implement: `contains`, `propertyNames`, `if|then|else`

**Comment by pateketrueke** [comment]
I would like to start exploring this enhancements but without breaking backwards, most people are stick to v4 and of course I don't plan to break too much their stuff.

There's other thing I really care about: reusability. It would be easier to reuse future-proof schemas right now?

I'm interested on projects like [react-jsonschema-form](https://github.com/mozilla-services/react-jsonschema-form) and [json-schema-sequelizer](https://github.com/pateketrueke/json-schema-sequelizer) which supports only v4 schemas.

For me, having the same basic schema and being able to reuse it on different tools is a win-win.

So, at this moment is not priority but certainly It'll be soon...

I would to start this just by listing which feature we plan to support in short/mid/long term, any guidance on this is highly appreciated.

Thank you for sharing your time & thoughts.

**Comment by robwalch** [comment]
I'd really like to see an option to use a random value from `examples` if present in the schema.

**Comment by pateketrueke** [comment]
> I'd really like to see an option to use a random value from examples if present in the schema.

Sounds like a good PR if you ask me. 👍

**Comment by robwalch** [comment]
@pateketrueke 👍 Yes!

**Comment by pateketrueke** [comment]
Related with #466

**Comment by ldexterldesign** [comment]
Hi @pateketrueke,

Thanks for software

Hope you're well

Your project seems very unique!

With >v4 schema support you'd make it onto https://json-schema.org/implementations and get a lot more exposure; even with just a few features, as @ehmicky asserts:

> Some aspects of JSON schema v6 and v7 are complex, but some might be low-hanging fruits. Libraries like ajv often only support parts of the most recent properties.

Hope this helps

Keep up great work!

Regards

**Comment by pateketrueke** [comment]
Thank you, definitely this is a MUST to achieve asap!

So much thanks for your feedback.

**Comment by philsturgeon** [comment]
@pateketrueke hey, I'd like to offer my help with getting this up to support draft 7 (6 is now outta date). 

For example, over at Stoplight.io we could do with readOnly/writeOnly support, but I don't want to start jamming in draft6/7 features when the baseline support is draft4.

**Comment by pateketrueke** [comment]
Hi @philsturgeon how those keywords would work?

**Comment by ehmicky** [comment]
I was wondering the same. [`readOnly` and `writeOnly`](https://json-schema.org/draft-07/json-schema-validation.html#rfc.section.10.3) are annotation/metadata keywords, not sure how this would apply to fake data generation.

**Comment by pateketrueke** [comment]
That's the question, keywords that are used for validation **can be** used also for data generation as they're counter parts.

Only if they affect how the data is being validated against the schema then we can implement something, otherwise I believe `readOnly/writeOnly` are not an issue yet.

Do you have some insights of how those keywords are validated?

**Comment by philsturgeon** [comment]
Take a look at https://github.com/stoplightio/prism/issues/524 dor a use-case.

A similar tool is [openapi-sampler](https://github.com/Redocly/openapi-sampler) which also has optional support readOnly/writeOnly. 

Basically it’ll ignore fields in a certain context, which is what we need for the prism issue above. JSON Schema is getting the keywords so at some point we’ll need to add support. :)

**Comment by equalspeterd** [comment]
Adding additional clarity (since I opened the ticket https://github.com/stoplightio/prism/issues/524:
When generating mock data,`writeOnly` properties are never included in a response. So, as a schema designer, I can define one object, but control behaviors within POST request bodies (for write operations) and GET response bodies (for read). So an example of an instance of a specific schema (to mock a GET response) should never include a `writeOnly` property.

Perhaps the most obvious example would be some shared secret (like a credential) where a new value can be set, but is never returned in a responses to a GET request.

**Comment by pateketrueke** [comment]
So it's possible to include/exclude properties based on those keywords?

**Comment by equalspeterd** [comment]
> So it's possible to include/exclude properties based on those keywords?

Yes, but based on verb's operating on the resource (which I would imagine may be outside the scope of this project). But when `faker` serializes a instance of the schema, should `faker` need to understand the context (e.g. creating a `write` operations vs. a `read` operation)?

**Comment by philsturgeon** [comment]
This chat about readOnly and writeOnly is particularly relevant: https://github.com/OAI/OpenAPI-Specification/issues/1622

Can we change the name of this from draft 6 to draft 2019-09 and just jump right over draft 7?

**Comment by pateketrueke** [comment]
> Can we change the name of this from draft 6 to draft 2019-09 and just jump right over draft 7?

Yes!

I'm reading the link you posted, thank you.

**Comment by XVincentX** [comment]
@pateketrueke Hey!

This is coming up again here :)

Is there anything I can do to help making this happen?

**Comment by reubano** [comment]
> This chat about readOnly and writeOnly is particularly relevant: [OAI/OpenAPI-Specification#1622](https://github.com/OAI/OpenAPI-Specification/issues/1622)
> 
> Can we change the name of this from draft 6 to draft 2019-09 and just jump right over draft 7?

I don't think skipping draft 7 is a good idea. Draft 2019-09 is a lot more complex than draft 7 and isn't as widely supported. I'm creating my current schemas using draft 7.

**Comment by philsturgeon** [comment]
Seeing as [OpenAPI v3.1 is now using JSON Schema 2020-12](https://apisyouwonthate.com/blog/openapi-v3-1-and-json-schema) I'd recommend getting to that version as a priority over older drafts. 

Very few people support 2019-09 but adoption of 2020-12 is coming along nicely thanks to the push from OpenAPI.

**Comment by Smolations** [comment]
@pateketrueke Looks like it's been a few years since anyone pushed on this. Just giving a friendly nudge to see if we can start moving the needle on some of the more current drafts. Would be nice to at least move up to the next major draft version... 😅 🤞 

If responding, I would be curious if the major concerns would be addressed at this point:
1. Would you consider 0.5.x as currently being "stable?" Infrequent updates would indicate that it is. 🤔 
2. Are there enough contributors at this point (as seen on the main README) that would be able to help with such an effort?

**Comment by pateketrueke** [comment]
Sadly I am pretty alone no this project, I'll try to make some room during this year but I cannot promise anything! I need to invest time on upgrading all the tooling, dependencies as needed prior adding more support, etc.
